### PR TITLE
feat(react-scheduler): add grouping by date

### PR DESCRIPTION
**Pull request by AryamnovEugeniy**
Feat #2236

### CODE REVIEW

**Comment by AryamnovEugeniy** [comment]
FPR

**Comment by AryamnovEugeniy** [comment]
FPR

File: packages/dx-react-scheduler-demos/src/demo-sources/scheduler-grouping/material-ui/group-by-dates.jsx [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+  Appointments,
+  AppointmentTooltip,
+  AppointmentForm,
+  DragDropProvider,
+  GroupingPanel,
+  WeekView,
+} from '@devexpress/dx-react-scheduler-material-ui';
+import { data as appointments } from '../../../demo-data/grouping';
+
+const groupByDate = viewName => viewName === 'Week';
```
Comment (MaximKudriavtsev) [inline packages/dx-react-scheduler-demos/src/demo-sources/scheduler-grouping/material-ui/group-by-dates.jsx]: `isWeekView` ?

File: packages/dx-react-scheduler/docs/guides/grouping.md [hunk 2]
```diff
@@ -24,6 +24,12 @@ A single appointment can appear in multiple groups if it is assigned to [multipl
 
 .embedded-demo({ "path": "scheduler-grouping/basic", "showThemeSelector": true })
 
+## Grouping by Date
+
+The example below shows how to group appointments by date and then by resource. To group by date, use the `GroupingState`'s `groupByDate` property. You can change this order by clicking "Group by Date First" above the Scheduler.
```
Comment (MaximKudriavtsev) [inline packages/dx-react-scheduler/docs/guides/grouping.md]: In the **following demo**, you can change this order by clicking "Group by Date First" above the Scheduler.

File: packages/dx-react-scheduler/src/plugins/grouping-panel.tsx [hunk 3]
```diff
@@ -49,6 +49,7 @@ class GroupingPanelBase extends React.PureComponent<GroupingPanelProps> {
                   ? HORIZONTAL_VIEW_LEFT_OFFSET
                   : VERTICAL_VIEW_LEFT_OFFSET,
                 }}
+                showHeaderForEveryDate={groupByDate && groupByDate(currentView && currentView.name)}
```
Comment (MaximKudriavtsev) [inline packages/dx-react-scheduler/src/plugins/grouping-panel.tsx]:
```js
groupByDate(currentView && currentView.name)

groupByDate(currentView?.name)
```

File: packages/dx-react-scheduler/src/plugins/integrated-grouping.tsx [hunk 4]
```diff
@@ -31,14 +33,18 @@ const getGroupingComputed = (
...
-  { allDayAppointments, grouping, resourcesToGroupBy, groups }: Getters,
-) => allDayAppointments &&
-  expandGroups(allDayAppointments, grouping, resourcesToGroupBy, groups);
+const getAllDayAppointmentsComputed = ({
+    allDayAppointments, grouping, resourcesToGroupBy,
+    groups, groupByDate, currentView, excludedDays,
+  }: Getters) => allDayAppointments &&
+  expandGroups(
+    allDayAppointments, grouping, resourcesToGroupBy,
+    groups, excludedDays, groupByDate(currentView && currentView.name),
```
Comment (MaximKudriavtsev) [inline packages/dx-react-scheduler/src/plugins/integrated-grouping.tsx]: `currentView?.name`

**Comment by AryamnovEugeniy** [comment]
FPR

**Comment by AryamnovEugeniy** [comment]
FPR

File: packages/dx-scheduler-core/src/plugins/common/utils.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+export const getDayScaleCellColSpan: PureComputed<
+  [ViewCell[][]], number
+> = (cellsData) => {
+  const firstDate = moment(cellsData[0][0].startDate);
+  let count = 1;
+  while (cellsData[0][count] && moment(cellsData[0][count].startDate).isSame(firstDate, 'day')) {
+    count += 1;
+  }
+  return count;
```
Comment (MaximKudriavtsev) [inline packages/dx-scheduler-core/src/plugins/common/utils.ts]: Should we create a moment instance for `firstDate`?

**Comment by AryamnovEugeniy** [comment]
FPR

File: packages/dx-react-scheduler/src/plugins/grouping-state.test.tsx [hunk 6]
```diff
@@ -67,4 +67,24 @@ describe('GroupingState', () => {
         resourceName: 'a',
       }]);
   });
+  it('should provide "groupByDate" getter', () => {
+    const tree = mount((
+      <PluginHost>
+        <GroupingState
+          groupByDate={(viewName) => {
+            if (viewName === 'day') {
```
Comment (ushkal) [inline packages/dx-react-scheduler/src/plugins/grouping-state.test.tsx]: `return viewName !== 'day'`

File: packages/dx-react-scheduler/docs/reference/grouping-state.md [hunk 7]
```diff
@@ -21,6 +21,7 @@ None
 Name | Type | Default | Description
 -----|------|---------|------------
 grouping? | Array&lt;[Grouping](#grouping)&gt; | | Specifies resources to group by.
+groupByDate? | (viewName: string) => boolean | `() => false` | Specifies whether appointments should be grouped by date in a specific view.
```
Comment (ushkal) [inline packages/dx-react-scheduler/docs/reference/grouping-state.md]: `groupByDate` sounds more like an action than query. We should consider name like `isGroupByDate` or similar

File: packages/dx-react-scheduler-material-ui/src/templates/layouts/horizontal-view-layout.jsx [hunk 8]
```diff
@@ -24,14 +24,26 @@ const styles = theme => ({
     minWidth: '100%',
     display: 'table',
   },
+  ordinaryBorderHeader: {
```
Comment (ushkal) [inline packages/dx-react-scheduler-material-ui/src/templates/layouts/horizontal-view-layout.jsx]: BorderHeader -> HeaderBorder

File: packages/dx-react-scheduler/docs/reference/grouping-panel.md [hunk 9]
```diff
@@ -76,6 +77,8 @@ Field | Type | Description
 group | [Group](#group) | The group the cell represents.
 colSpan | number | The number of columns the cell spans.
 left | number | The cell's offset from the left.
+brightBorderRight? | boolean | Indicates whether to draw a bright right border or an ordinary right border.
```
Comment (ushkal) [inline packages/dx-react-scheduler/docs/reference/grouping-panel.md]: Though `brightRightBorder` sounds slightly unnatural, I think this name better suits here as it consistent with other props such as `hasTightBorder`

File: packages/dx-react-scheduler/docs/guides/grouping.md [hunk 10]
```diff
@@ -24,6 +24,12 @@ A single appointment can appear in multiple groups if it is assigned to [multipl
 
 .embedded-demo({ "path": "scheduler-grouping/basic", "showThemeSelector": true })
 
+## Grouping by Date
```
Comment (RomanTsukanov) [inline packages/dx-react-scheduler/docs/guides/grouping.md]:
```suggestion
## Group by Date
```

File: packages/dx-react-scheduler/docs/guides/grouping.md [hunk 11]
```diff
@@ -24,6 +24,12 @@ A single appointment can appear in multiple groups if it is assigned to [multipl
...
+
+The example below shows how to group appointments by date and then by resource. To group by date, use the `GroupingState`'s `groupByDate` property. In the following demo, you can change this order by clicking "Group by Date First" above the Scheduler.
```
Comment (RomanTsukanov) [inline packages/dx-react-scheduler/docs/guides/grouping.md]:
```suggestion
To group appointments by date, set the `GroupingState`'s `groupByDate` property to `true`.

In the example below, appointments are grouped first by date and then by resource. Click the "Group by Date First" switch to change this order.
```

File: packages/dx-react-scheduler/docs/reference/all-day-panel.md [hunk 12]
```diff
@@ -45,6 +45,8 @@ Field | Type | Description
 ------|------|------------
 startDate | Date | The cell's start time.
 endDate | Date | The cell's end time.
+groupingInfo? | Array&lt;[Group](./grouping-panel.md/#group)&gt; | Information about the cell's group(s).
+endOfGroup? | boolean | Specifies whether it's the last cell in a group.
```
Comment (RomanTsukanov) [inline packages/dx-react-scheduler/docs/reference/all-day-panel.md]:
```suggestion
endOfGroup? | boolean | `true` if this cell is last in its group.
```

File: packages/dx-react-scheduler/docs/reference/all-day-panel.md [hunk 13]
```diff
@@ -45,6 +45,8 @@ Field | Type | Description
 ------|------|------------
 startDate | Date | The cell's start time.
 endDate | Date | The cell's end time.
+groupingInfo? | Array&lt;[Group](./grouping-panel.md/#group)&gt; | Information about the cell's group(s).
```
Comment (RomanTsukanov) [inline packages/dx-react-scheduler/docs/reference/all-day-panel.md]:
```suggestion
groupingInfo? | Array&lt;[Group](./grouping-panel.md/#group)&gt; | Information about the cell's grouping.
```

File: packages/dx-react-scheduler/docs/reference/day-view.md [hunk 14]
```diff
@@ -97,6 +99,7 @@ cellsData | Array&lt;Array&lt;[DayView.CellData](#dayviewcelldata)&gt;&gt; | Spe
 cellComponent | ComponentType&lt;[DayView.DayScaleCellProps](#dayviewdayscalecellprops)&gt; | A component that renders a day scale cell.
 rowComponent | ComponentType&lt;[DayView.RowProps](#dayviewrowprops)&gt; | A component that renders a day scale row.
 groupingPanelComponent? | ComponentType&lt;[GroupingPanel.HorizontalLayoutProps](./grouping-panel.md/#groupingpanelhorizontallayoutprops)&gt; | A component that renders the grouping panel.
+groupedByDate? | boolean | Indicates whether grouping by date is enabled.
```
Comment (RomanTsukanov) [inline packages/dx-react-scheduler/docs/reference/day-view.md]:
```suggestion
groupedByDate? | boolean | Indicates whether grouping by date is enabled. Takes its value from the `groupByDate` property of the [GroupingState]() plugin.
```

Comment (RomanTsukanov) [reply packages/dx-react-scheduler/docs/reference/day-view.md]: Please add a link to `GroupingState` and make the same change to every `groupedByDate` description.

File: packages/dx-react-scheduler/docs/reference/all-day-panel.md (hunk 12 above)
Comment (RomanTsukanov) [reply packages/dx-react-scheduler/docs/reference/all-day-panel.md]: Please make the same change for every `endOfGroup` description.

File: packages/dx-react-scheduler/docs/reference/all-day-panel.md (hunk 13 above)
Comment (RomanTsukanov) [reply packages/dx-react-scheduler/docs/reference/all-day-panel.md]: Please make the same change for every `groupingInfo` description that involves cell.

**Pull request merged**